### PR TITLE
Problem: parallel transaction execution is not supported

### LINKED ISSUE #432: [EPIC] block-stm integration
- [x] https://github.com/crypto-org-chain/cosmos-sdk/pull/206
- [x] https://github.com/crypto-org-chain/ethermint/pull/447
- [x] https://github.com/crypto-org-chain/ethermint/pull/448
- [x] https://github.com/crypto-org-chain/go-block-stm/pull/5
- [ ] Optimize `Has` operation in go-block-stm
- [x] https://github.com/crypto-org-chain/ethermint/pull/456
- [x] https://github.com/crypto-org-chain/ethermint/pull/450

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #432

## Description

Integrate go-block-stm to do parallel tx execution.

- refactored the transient stores to be tx-index based, so the tx execution don't conflicts
- add a patch call after all txs executed to decide the tx index/log size stuff.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,6 +39,13 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### Features
 
+* (app) [#407](https://github.com/crypto-org-chain/ethermint/pull/407) Integrate sdk 0.50.x.
+* (evm) [#]() Integrate go-block-stm for parallel tx execution.
```
Comment (yihuang) [inline CHANGELOG.md]:
```suggestion
* (evm) [#414](https://github.com/crypto-org-chain/ethermint/pull/414) Integrate go-block-stm for parallel tx execution.
```

File: app/executor.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+type msWrapper struct {
+	block_stm.MultiStore
+	stores     []storetypes.StoreKey
+	keysByName map[string]storetypes.StoreKey
+}
+
+var _ storetypes.MultiStore = msWrapper{}
+
+func newMultiStoreWrapper(ms block_stm.MultiStore, stores []storetypes.StoreKey) msWrapper {
+	keysByName := make(map[string]storetypes.StoreKey)
```
Comment (yihuang) [inline app/executor.go]:
```suggestion
	keysByName := make(map[string]storetypes.StoreKey, len(stores))
```

**Review by mmsqe — APPROVED** [review]

**Comment by yihuang** [comment]
put to draft for now until we get the sdk ready.

File: app/executor.go [hunk 3]
```diff
@@ -19,3 +25,129 @@ func DefaultTxExecutor(_ context.Context,
...
+	panic("cannot branch cached multi-store with a version")
+}
+
+// Implements CacheWrapper.
+func (ms msWrapper) CacheWrap() storetypes.CacheWrap {
+	return ms.CacheMultiStore().(storetypes.CacheWrap)
+}
+
+// CacheWrapWithTrace implements the CacheWrapper interface.
+func (ms msWrapper) CacheWrapWithTrace(_ io.Writer, _ storetypes.TraceContext) storetypes.CacheWrap {
```
Comment (mmsqe) [inline app/executor.go]: seems no need CacheWrapWithTrace now

Comment (yihuang) [reply app/executor.go]: good catch

File: x/evm/types/utils.go [hunk 4]
```diff
@@ -204,3 +204,11 @@ func GetBaseFee(height int64, ethCfg *params.ChainConfig, feemarketParams *feema
 	}
 	return baseFee
 }
+
+// CopyAppend returns a new slice with all the elements of a followed by all the elements of b.
+func CopyAppend(a, b []byte) []byte {
```
Comment (mmsqe) [inline x/evm/types/utils.go]: seems no need now

File: x/evm/types/interfaces.go [hunk 5]
```diff
@@ -46,6 +46,7 @@ type AccountKeeper interface {
 type BankKeeper interface {
 	authtypes.BankKeeper
 	GetBalance(ctx context.Context, addr sdk.AccAddress, denom string) sdk.Coin
+	GetAllBalances(ctx context.Context, addr sdk.AccAddress) sdk.Coins
```
Comment (mmsqe) [inline x/evm/types/interfaces.go]: do we still need this?

File: x/evm/keeper/abci.go [hunk 6]
```diff
@@ -35,8 +33,6 @@ func (k *Keeper) BeginBlock(ctx sdk.Context) error {
 // KVStore. The EVM end block logic doesn't update the validator set, thus it returns
 // an empty slice.
 func (k *Keeper) EndBlock(ctx sdk.Context) error {
```
Comment (mmsqe) [inline x/evm/keeper/abci.go:35]: do we still need [reset](https://github.com/crypto-org-chain/ethermint/blob/a5b1f1c44016976bb14da255ac7424d3f1199323/x/evm/keeper/state_transition.go#L245-L251) the gas meter?

Comment (yihuang) [reply x/evm/keeper/abci.go:35]: that one is needed, that's how the cosmos tx calculate the gas used

File: x/evm/types/interfaces.go (hunk 5 above)
Comment (yihuang) [reply x/evm/types/interfaces.go]: nop, good catch

File: x/evm/types/utils.go (hunk 4 above)
Comment (yihuang) [reply x/evm/types/utils.go]: yes, removed

**Review by mmsqe — APPROVED** [review]

File: server/config/config.go [hunk 7]
```diff
@@ -124,6 +124,10 @@ type EVMConfig struct {
 	Tracer string `mapstructure:"tracer"`
 	// MaxTxGasWanted defines the gas wanted for each eth tx returned in ante handler in check tx mode.
 	MaxTxGasWanted uint64 `mapstructure:"max-tx-gas-wanted"`
+	// BlockExecutor set block executor type, "block-stm" for parallel execution, "default" for sequential execution.
```
Comment (yihuang) [inline server/config/config.go]:
```suggestion
	// BlockExecutor set block executor type, "block-stm" for parallel execution, "sequential" for sequential execution.
```

File: server/config/toml.go [hunk 8]
```diff
@@ -31,6 +31,11 @@ tracer = "{{ .EVM.Tracer }}"
 # MaxTxGasWanted defines the gas wanted for each eth tx returned in ante handler in check tx mode.
 max-tx-gas-wanted = {{ .EVM.MaxTxGasWanted }}
 
+# BlockExecutor set block executor type, "block-stm" for parallel execution, "default" for sequential execution.
```
Comment (yihuang) [inline server/config/toml.go]:
```suggestion
# BlockExecutor set block executor type, "block-stm" for parallel execution, "sequential" for sequential execution.
```

**Pull request merged**